api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/generic/AgeOnObsDatetimeCalculation.java

```diff
@@ -0,0 +1,145 @@
+import org.openmrs.module.eptsreports.reporting.utils.EptsCalculationUtils;
+import org.openmrs.module.reporting.common.Birthdate;
+import org.openmrs.module.reporting.data.patient.definition.SqlPatientDataDefinition;
+import org.openmrs.module.reporting.data.person.definition.BirthdateDataDefinition;
+import org.openmrs.module.reporting.evaluation.parameter.Parameter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * The Class calculates the age of the patient based on report obs_datetime <br>
+ * <code>PatientAge = obs_datetime - PatientBirthDate</code> <br>
+ * The <i>minAge</i> and the <i>maxAge</i> are the boundaries to evaluate if the patient belong to
+ * specific renge of age
+ */
+@Component
+public class AgeOnObsDatetimeCalculation extends AbstractPatientCalculation {
+  private static final String MAX_AGE = "maxAgeOnObsDatetime";
+
+  private static final String MIN_AGE = "minAgeOnObsDatetime";
+
+  private final String ON_OR_BEFORE = "onOrBefore";
+
+  private final String ON_OR_AFTER = "onOrAfter";
+
+  private final String LOCATION = "location";
+
+  @Autowired private HivMetadata hivMetadata;
+
+  @Autowired private CommonMetadata commonMetadata;
+
+  @Override
+  public CalculationResultMap evaluate(
+      Collection<Integer> cohort,
+      Map<String, Object> parameterValues,
+      PatientCalculationContext context) {
+
+    CalculationResultMap map = new CalculationResultMap();
+
+    CalculationResultMap birthDates =
+        EptsCalculationUtils.evaluateWithReporting(
+            new BirthdateDataDefinition(), cohort, null, null, context);
+
+    CalculationResultMap PatientAndObsDatetimeMap = getPatientAgeObsDatetimeMap(cohort, context);
+
+    Integer minAge = (Integer) parameterValues.get(MIN_AGE);
+    Integer maxAge = (Integer) parameterValues.get(MAX_AGE);
+
+    for (Integer patientId : cohort) {
+      Date birthDate = getBirthDate(patientId, birthDates);
+      Date patientAndObsDatetime =
+          EptsCalculationUtils.resultForPatient(PatientAndObsDatetimeMap, patientId);
+
+      if (patientAndObsDatetime != null && birthDate != null) {
+        final boolean datesConsistent = patientAndObsDatetime.compareTo(birthDate) >= 0;
+        if (datesConsistent) {
+          int years =
+              Years.yearsIn(new Interval(patientAndObsDatetime.getTime(), birthDate.getTime()))
+                  .getYears();
+          boolean b = isMinAgeOk(minAge, years) && isMaxAgeOk(maxAge, years);
+          map.put(patientId, new BooleanResult(b, this));
+        }
+      }
+    }
+    return map;
+  }
+
+  private boolean isMaxAgeOk(Integer maxAge, int years) {
+    return maxAge == null || years <= maxAge.intValue();
+  }
+
+  private boolean isMinAgeOk(Integer minAge, int years) {
+    return minAge == null || years >= minAge.intValue();
+  }
+
+  private CalculationResultMap getPatientAgeObsDatetimeMap(
```

Comment:
The name of this function is not suggestive +put some javadocs before any function and class